### PR TITLE
fix: upscale not working on tiles with partial transparancy

### PR DESCRIPTION
**Pull request by vitalymelni**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✔                                                                        |
| New feature     | ✖                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✖                                                                       |

Tiles with partial transparency were treated as fully transparent and ignored when upscaled
This PR fixes the check in the upscale method if the tile is fully transparent correctly

### CODE REVIEW

File: MergerLogic/Utils/ImageUtils.cs [hunk 1]
```diff
@@ -6,15 +6,28 @@ public class ImageUtils
...
         public static bool IsTransparent(MagickImage image)
         {
-            if(!image.HasAlpha) {
+            if (!image.HasAlpha)
+            {
                 return false;
             }
 
             using var pixels = image.GetPixels();
             // Check pixels to see if all are transparent
```
Comment (shimoncohen) [inline MergerLogic/Utils/ImageUtils.cs]: Could you please fix the comment? It should be `if at least one pixel is transparent`.

File: MergerLogicUnitTests/ImageProcessing/TileScalerTest.cs [hunk 2]
```diff
@@ -42,6 +42,13 @@ public void BeforeEach()
 
         public static IEnumerable<object[]> GetUpscaleTilesTestParameters()
         {
+            yield return new object[] {
```
Comment (shimoncohen) [inline MergerLogicUnitTests/ImageProcessing/TileScalerTest.cs:44]: Could you please add a test where the upscaled tile is fully transparent and is discarded (null tile)?

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerLogic/Utils/ImageUtils.cs (hunk 1 above)
Comment (vitalymelni) [reply MergerLogic/Utils/ImageUtils.cs]: fixed

File: MergerLogicUnitTests/ImageProcessing/TileScalerTest.cs (hunk 2 above)
Comment (vitalymelni) [reply MergerLogicUnitTests/ImageProcessing/TileScalerTest.cs:44]: added

File: MergerLogic/Utils/ImageUtils.cs [hunk 3]
```diff
@@ -6,15 +6,28 @@ public class ImageUtils
     {
         public static bool IsTransparent(MagickImage image)
```
Comment (asafMasa) [inline MergerLogic/Utils/ImageUtils.cs:7]: add unit tests (full, partial, transparent)

File: MergerLogic/Utils/ImageUtils.cs [hunk 4]
```diff
@@ -6,15 +6,28 @@ public class ImageUtils
...
             }
 
             using var pixels = image.GetPixels();
-            // Check pixels to see if all are transparent
+            // Check pixels to see if at least one of them is transparent (or partially transparent)
             return pixels.Select(pixel => pixel.ToColor()).Any(color => color?.A != 255);
         }
-        
+
+        public static bool IsFullyTransparent(MagickImage image)
```
Comment (asafMasa) [inline MergerLogic/Utils/ImageUtils.cs:19]: add unit tests (full, partial, transparent)

**Review by asafMasa — CHANGES_REQUESTED** [review]
Good job!

File: MergerLogic/Utils/ImageUtils.cs (hunk 3 above)
Comment (vitalymelni) [reply MergerLogic/Utils/ImageUtils.cs:7]: added

File: MergerLogic/Utils/ImageUtils.cs (hunk 4 above)
Comment (vitalymelni) [reply MergerLogic/Utils/ImageUtils.cs:19]: added

File: MergerLogicUnitTests/Utils/ImageUtilsTest.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        true,
+      };
+
+      yield return new object[] {
+        File.ReadAllBytes("partial_transparent.png"),
+        true,
+      };
+
+      yield return new object[] {
+        File.ReadAllBytes("no_transparency.jpeg"),
```
Comment (asafMasa) [inline MergerLogicUnitTests/Utils/ImageUtilsTest.cs:36]: Add opaque png test

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: MergerLogicUnitTests/Utils/ImageUtilsTest.cs (hunk 5 above)
Comment (vitalymelni) [reply MergerLogicUnitTests/Utils/ImageUtilsTest.cs:36]: done

**Review by asafMasa — APPROVED** [review]

File: MergerLogicUnitTests/ImageProcessing/TileScalerTest.cs [hunk 6]
```diff
@@ -90,17 +89,39 @@ public static IEnumerable<object[]> GetUpscaleTilesTestParameters()
...
         [DynamicData(nameof(GetUpscaleTilesTestParameters), DynamicDataSourceType.Method)]
-        public void Upscale(byte[] tileBytes, Coord baseTileCoord, Coord targetCoord, byte[] expectedTileBytes)
+        public void Upscale(byte[] tileBytes, Coord baseTileCoord, Coord targetCoord, byte[]? expectedTileBytes)
         {
             var testTile = new Tile(baseTileCoord, tileBytes);
             var resultTile = this._testTileScaler.Upscale(testTile, targetCoord);
 
-            Assert.IsNotNull(resultTile);
-            CollectionAssert.AreEqual(expectedTileBytes, resultTile.GetImageBytes());
+            if (expectedTileBytes is null)
```
Comment (shimoncohen) [inline MergerLogicUnitTests/ImageProcessing/TileScalerTest.cs:116]: I'm not sure this is the best approach, maybe we should add a parameter indicating we expect the result to be `null`. This can happen when we get a `PNG` image too.

File: MergerLogicUnitTests/Utils/TestData/no_transparency.png
Comment (shimoncohen) [inline MergerLogicUnitTests/Utils/TestData/no_transparency.png:1]: How was this file created? It looks like it wasn't merged correctly...

**Review by shimoncohen — CHANGES_REQUESTED** [review]
